### PR TITLE
Feature: Add edit set annotation

### LINKED ISSUE #4051: [Question] Change annotations transformers value on fly with kustomize edit set
Hi,

first, thank you for this great and useful project. Second, I have a question I could not find an answer on.

Initially, I wanted to put some annotations to my `Deployment` resource. I used `commonAnnotations` setting in `kustomization.yml`. When I built it, I found it is applied to all resources. After, I found that I can solve it by built-in plugins and using annotations transformers. I tried this and succeed, it changes only the `Deployment`resource.

I have done with the following annotations transformer:

```
apiVersion: builtin
kind: AnnotationsTransformer
metadata:
  name: project-name
annotations:
  project-annotation: annotation
fieldSpecs:
- kind: Deployment
  path: metadata/annotations
  create: true
```

I was led by [this example](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/configureBuiltinPlugin.md), and it completely worked for me.

The only problem I have here is I want to be able to change the annotation value (`project-annotation: annotation`) to any other as I do with an image tag `kustomize edit set image project=2.4.1`.

I can do it with `sed` CLI, but was wondering if I could do it with `kustomize` CLI?

### PR DESCRIPTION
**Pull request by adberger**
This PR adds the `edit set annotation` to kustomize

Most code is copied from `kustomize/commands/edit/set/setlabel.go` as suggested.
Improved testing of annotation key validity as described in https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/

resolves #4051

### CODE REVIEW

**Comment by natasha41575** [comment]
ugh, I can't spell
/ok-to-test

File: kustomize/commands/edit/set/setannotation.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation.go]:
```suggestion
		if !IsValidKey(key) {
```

File: kustomize/commands/edit/set/setannotation.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation.go]: please add a link to the documentation for valid annotation keys, e.g.
```
// see https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
```

File: kustomize/commands/edit/set/setannotation.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation.go:25]: if we're going to do this validation, can we also add to the regexp that the length of the prefix must be {0,253} chars and length of the name must be {0,63} chars? This isn't a big deal and we can do without it but it's simple enough to add I think

File: kustomize/commands/edit/set/setannotation.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation.go:79]: Please remove the print statement :)

File: kustomize/commands/edit/set/setannotation.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation.go:80]: Let's change the error message to something like:
`invalid annotation key: see the syntax and character set rules at https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/`

**Review by natasha41575 — CHANGES_REQUESTED** [review]

File: kustomize/commands/edit/set/setannotation.go
Comment (adberger) [reply kustomize/commands/edit/set/setannotation.go]: ugh very bad, thanks!

File: kustomize/commands/edit/set/setannotation.go
Comment (adberger) [reply kustomize/commands/edit/set/setannotation.go]: Please check the new regex carefully :)

**Comment by adberger** [comment]
@natasha41575 Do I need to do something in addition for another review from yours?

**Comment by natasha41575** [comment]
I think we need some more tests for the regex. For example, running `kustomize edit set annotation internal.config.kubernetes.io/generatorOptions:true` currently fails (as does anything else I try with a prefix). Please add tests for

1. The above example, `kustomize edit set annotation internal.config.kubernetes.io/generatorOptions:true`
2. An annotation that has a key with prefix 253 characters and a name with 63 characters.
3. An annotation that has a key with prefix 252 characters and a name with 62 characters.
4. An annotation that has a key with prefix 254 characters and a name with 64 characters.

**Comment by adberger** [comment]
I fixed the regex as follows:
* Added `A-Z` for prefix
* Added `.` in prefix (but not start & end)
* Removed numerical characters `0-9` from start of prefix

I tried my best to extend the tests as you suggested, please check them and tell me if it's ok now.

File: kustomize/commands/edit/set/setannotation_test.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation_test.go:101]: instead of using backticks (` `` `), let's do 

```
"abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstu" +
		"vwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz" +
		"abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcde" +
		"fghijklmnopqrstuvwxyzabcdefghijklmnopqrs/abcdefghijklmnopqrstuvwxyzabcdefghijklmnop" +
		"qrstuvwxyzabcdefghijk"
```

It seems a bit weird to have newlines in an annotation key (and this test also fails for me locally with backticks).

File: kustomize/commands/edit/set/setannotation_test.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation_test.go:118]: same thing here, let's do 
```
"abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghi" +
		"jklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmn" +
		"opqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrs" +
		"tuvwxyzabcdefghijklmnopqrst/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabc" +
		"defghij:true"
```
instead of backticks to avoid the newlines.

File: kustomize/commands/edit/set/setannotation_test.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation_test.go:185]: add `t.FailNow()` after this

File: kustomize/commands/edit/set/setannotation_test.go
Comment (natasha41575) [inline kustomize/commands/edit/set/setannotation_test.go:122]: ditto, add `t.FailNow()` after this. This is to avoid panicking in line 124 if `err==nil`

**Review by natasha41575 — COMMENTED** [review]
Looks pretty good! Just a couple of minor nits and we should be able to merge this soon. Thank you!

File: kustomize/commands/edit/set/setannotation_test.go
Comment (adberger) [reply kustomize/commands/edit/set/setannotation_test.go:190]: This seemed a bit odd to add this in only 2 places when there were 6 places which have `err==nil` condition.
Please let me know if all 6 need `t.FailNow()` or only the 2 you specified.
If only the 2, could you elaborate on why this is only needed there?

Comment (natasha41575) [reply kustomize/commands/edit/set/setannotation_test.go:190]: Yes, all 6 of them

**Comment by natasha41575** [comment]
Tests look good; I'll pull the branch when I get some time to do some manual local testing as well.

**Comment by natasha41575** [comment]
This is almost there. Most cases such as `kustomize edit set internal.config.kubernetes.io/options: true` seem to be working as expected. 

But I find something very strange when running this locally:

```
$ kustomize edit set annotation abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkl/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijk:value
```

results in the following appearing in my kustomization file: 
```
commonAnnotations:
  ? abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkl/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijk
  : value
```

Any ideas why the question mark is showing up?

**Comment by adberger** [comment]
@natasha41575 The "bug" not only exists for the `set annotation` command, try `set label` for example.
I tracked the issue down to the `JSONToYAML` function from the go yaml package (v1.2.0).

kf.Write(m)
-> data, err := mf.marshal(kustomization)
---> content, err := marshalField(field, kustomization)
-----> return yaml.Marshal(k)
-------->y, err := JSONToYAML(j)

https://imgur.com/a/4LiqB1v

`Println(t)` the output seems still correct
`Println(s)` the output is wrong

The problematic part seems to be somewhere in the `Marshal` function of the yaml.v2 package (v2.4.0):
```
func Marshal(in interface{}) (out []byte, err error) {
	defer handleErr(&err)
	e := newEncoder()
	defer e.destroy()
	e.marshalDoc("", reflect.ValueOf(in))
	e.finish()
	out = e.out
	return
}
```

My go knowledge isn't enough to really understand where the problem is, maybe someone can help me on this one.

**Comment by natasha41575** [comment]
> The "bug" not only exists for the set annotation command, try set label for example.
I tracked the issue down to the JSONToYAML function from the go yaml package (v1.2.0).

Weird. Since it also happens for `set label` I don't see a reason to hold this PR, but if you can, I would appreciate if you could file a follow-up issue with the same contents/explanation that you have in your above comment.

/lgtm
/approve

**Review by natasha41575 — APPROVED** [review]

**Pull request merged**
